### PR TITLE
Feat/slot group

### PR DESCRIPTION
**Pull request by wenpq**
- A button to add slots to specified slot group.
- A button to remove slots from specified slot group.
- Basic slot group model.
- Generate slots table data for testing.  
- Modifications on slot model.

### CODE REVIEW

File: models/slot-group.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+var mongoose = require('mongoose');
+var Schema = mongoose.Schema;
+var ObjectId = Schema.Types.ObjectId;
+var Mixed = Schema.Types.Mixed;
+
+var slotGroup = new Schema({
+  name: {
+    type: String,
+    index: {
```
Comment (dongliu) [inline models/slot-group.js]: I am not sure the syntax is correct. just use {index: true, unique: true}
See https://github.com/Automattic/mongoose/issues/2232

File: models/slot.js [hunk 2]
```diff
@@ -5,25 +5,23 @@ var ObjectId = Schema.Types.ObjectId;
 var slot = new Schema({
   system: String,
   subsystem: String,
-  deviceNaming: String,// mapping to device column in slot excel file
+  deviceNaming: String,// mapping to 'Device' column in slot excel file
   beamlinePosition: Number,// Beam line position (dm)
-  name: String,
+  name: {
+    type: String,
+    index: {
```
Comment (dongliu) [inline models/slot.js]: need to update the syntax

File: views/slot-group.jade [hunk 3]
```diff
@@ -0,0 +1,10 @@
+extends layout-history
+
+block head
+  link(rel='stylesheet', href='/datatables/media/css/dataTables.bootstrap.min.css')
+  title RunCheck device
```
Comment (dongliu) [inline views/slot-group.jade]: title should be slot groups

File: models/slot-group.js (hunk 1 above)
Comment (wenpq) [inline models/slot-group.js]: You are right, I will correct it.

File: routes/slots.js [hunk 4]
```diff
@@ -1,34 +1,214 @@
...
+        res.status(200).json({
+          passDataId: passDataId,
+          conflictDataName: conflictDataName
+        });
+      }
+    });
+  });
+});
+
+slots.post('/addGroup',auth.ensureAuthenticated, function (req, res) {
```
Comment (dongliu) [inline routes/slots.js]: the resource to handle this should be /slotgroups/:id

File: routes/slots.js [hunk 5]
```diff
@@ -1,34 +1,214 @@
...
+  // change slots of slotGroup
+  SlotGroup.findOne({'name': req.body.slotGroupName}, function(err, slotGroup){
+    if(err) {
+      console.error(err);
+      return res.status(500).send(err.message);
+    }
+    slotGroup.slots = slotGroup.slots.concat(req.body.slotIds);// concat array
```
Comment (dongliu) [inline routes/slots.js]: we need to do validation here. check if a slot id exists and the slot.inGroup is undefined.

File: routes/slots.js [hunk 6]
```diff
@@ -1,34 +1,214 @@
...
+        console.error(err);
+        return res.status(500).send(err.message);
+      }
+      // change inGroup of slot
+      Slot.update({ '_id': {$in: req.body.slotIds}}, {inGroup: slotGroup._id}, {multi: true}, function(err) {
+        if(err) {
+          console.error(err);
+          return res.status(500).send(err.message);
+        }
+        res.status(200).end();
```
Comment (dongliu) [inline routes/slots.js]: response with the list of slots in this slot group

File: routes/slots.js (hunk 5 above)
Comment (dongliu) [inline routes/slots.js]: need to use add to set methods see https://docs.mongodb.com/manual/reference/operator/update/addToSet/

File: routes/slots.js [hunk 7]
```diff
@@ -1,34 +1,214 @@
...
+      })
+    });
+  });
+});
+
+slots.post('/removeGroup',auth.ensureAuthenticated, function (req, res) {
```
Comment (dongliu) [inline routes/slots.js]: similarly, this should be in /slotgroups/:id resource route. And we might want to remove by DELETE /slotgroups/:gid/slots/:sid

File: routes/slots.js [hunk 8]
```diff
@@ -1,34 +1,214 @@
...
+    }
+    // delete specified slots of slotGroup
+    SlotGroup.find(function(err, docs){
+      if(err) {
+        console.error(err);
+        return res.status(500).send(err.message);
+      }
+      var count = 0;
+      docs.forEach(function(slotGroup) {
+        slotGroup.slots = slotGroup.slots.filter(function (x) {
```
Comment (dongliu) [inline routes/slots.js]: use pull to remove from the set see https://docs.mongodb.com/manual/reference/operator/update/pull/

**Pull request merged**